pkg/controllers/inventory/inventory_controller.go

```diff
@@ -0,0 +1,168 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/config"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	commonservice "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/inventory"
+)
+
+const (
+
+	// How long to wait before retrying
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+
+	inventoryGCJitterFactor = 0.1
+)
+
+type WatchResourceFunc func(c *InventoryController, mgr ctrl.Manager) error
+
+var (
+	log = &logger.Log
+
+	// DeletionHandlingMetaNamespaceKeyFunc checks for
+	// DeletedFinalStateUnknown objects before calling
+	// MetaNamespaceKeyFunc.
+	keyFunc            = cache.DeletionHandlingMetaNamespaceKeyFunc
+	WatchResourceFuncs = []WatchResourceFunc{watchPod}
+)
+
+type InventoryController struct {
+	Client               client.Client
+	service              *inventory.InventoryService
+	inventoryObjectQueue workqueue.TypedRateLimitingInterface[any]
+	keyBuffer            sets.Set[inventory.InventoryKey]
+	inventoryMutex       sync.Mutex
+	cf                   *config.NSXOperatorConfig
+}
+
+func NewInventoryController(Client client.Client, service *inventory.InventoryService, cf *config.NSXOperatorConfig) *InventoryController {
+	limiter := workqueue.NewTypedItemExponentialFailureRateLimiter[any](minRetryDelay, maxRetryDelay)
+	queueConfig := workqueue.TypedRateLimitingQueueConfig[any]{Name: "inventoryObject"}
+	queue := workqueue.NewTypedRateLimitingQueueWithConfig(limiter, queueConfig)
+	c := &InventoryController{
+		Client:               Client,
+		service:              service,
+		keyBuffer:            sets.New[inventory.InventoryKey](),
+		cf:                   cf,
+		inventoryObjectQueue: queue,
+	}
+	return c
+}
+
+func StartupInventoryController(mgr ctrl.Manager, service *inventory.InventoryService, cf *config.NSXOperatorConfig) error {
+	controller := NewInventoryController(mgr.GetClient(), service, cf)
+	return controller.SetupWithManager(mgr)
+}
+
+func (c *InventoryController) SetupWithManager(mgr ctrl.Manager) error {
+	for _, f := range WatchResourceFuncs {
+		err := f(c, mgr)
+		if err != nil {
+			return err
+		}
+	}
+	// Set up the queue
+	go c.Run(make(<-chan struct{}))
+	return nil
+}
+
+func (c *InventoryController) Run(stopCh <-chan struct{}) {
+	defer c.inventoryObjectQueue.ShutDown()
+	log.Info("Starting inventory controller")
+
+	log.Info("Caches are synced for inventory controller, cleaning up stale objects")
+	err := c.CleanStaleInventoryObjects()
+	if err != nil {
+		log.Error(err, "Failed to clean up stale inventory objects")
+		return
+	}
+
+	// Batch update inventory based on batch time and size.
+	// Inventory worker will be running in forever loop util inventoryMutex is locked by inventoryTimeWorker.
```

Comment:
The inventory worker will operate in an endless loop, unless interrupted, with the inventoryMutex locked by the inventoryTimeWorker.

Reply:
util -> until ?

Reply:
Done